overlord/snapstate/snapstate_test.go

```diff
@@ -5314,6 +5314,138 @@ func (s *snapmgrTestSuite) TestRemoveMany(c *C) {
+		case 1:
+			c.Check(required, Equals, snapstate.SafetyMarginDiskSpace(10))
+		case 2:
+			c.Check(required, Equals, snapstate.SafetyMarginDiskSpace(20))
+		case 3:
+			c.Check(required, Equals, snapstate.SafetyMarginDiskSpace(30))
+		default:
+			c.Fatalf("unexpected osutil.CheckFreeSpace call")
+		}
+		if failingDiskSizeCheck == checkFreeSpaceCall {
+			return &osutil.NotEnoughDiskSpaceError{}
+		}
+		return nil
+	})
+	defer restore()
+
+	var automaticSnapshotCalled bool
+	snapstate.AutomaticSnapshot = func(st *state.State, instanceName string) (ts *state.TaskSet, err error) {
+		automaticSnapshotCalled = true
+		if automaticSnapshot {
+			t := s.state.NewTask("foo", "")
+			ts = state.NewTaskSet(t)
+			return ts, nil
+		}
+		// ErrNothingToDo is returned if automatic snapshots are disabled
+		return nil, snapstate.ErrNothingToDo
+	}
+
+	tr := config.NewTransaction(s.state)
+	tr.Set("core", "experimental.check-disk-space-remove", featureFlag)
+	tr.Commit()
+
+	snapstate.Set(s.state, "one", &snapstate.SnapState{
+		Active:   true,
+		SnapType: "app",
+		Sequence: []*snap.SideInfo{
+			{RealName: "one", SnapID: "one-id", Revision: snap.R(1)},
+		},
+		Current: snap.R(1),
+	})
+	snapstate.Set(s.state, "two", &snapstate.SnapState{
+		Active:   true,
+		SnapType: "app",
+		Sequence: []*snap.SideInfo{
+			{RealName: "two", SnapID: "two-id", Revision: snap.R(1)},
+		},
+		Current: snap.R(1),
+	})
+
+	_, _, err := snapstate.RemoveMany(s.state, []string{"one", "two"})
+
+	if featureFlag && automaticSnapshot {
+		c.Check(checkFreeSpaceCall, Equals, failingDiskSizeCheck)
+		c.Check(snapshotSizeCall, Equals, 2)
+		c.Check(automaticSnapshotCalled, Equals, true)
+	} else {
+		c.Check(checkFreeSpaceCall, Equals, 0)
+		c.Check(snapshotSizeCall, Equals, 0)
+	}
+
+	c.Check(automaticSnapshotCalled, Equals, true)
+
+	return err
+}
+
+func (s *snapmgrTestSuite) TestRemoveManyDiskSpaceError(c *C) {
+	featureFlag := true
+	automaticSnapshot := true
+	failingDiskCheck := 3
+	err := s.testRemoveManyDiskSpaceCheck(c, featureFlag, automaticSnapshot, failingDiskCheck)
+
+	diskSpaceErr := err.(*snapstate.InsufficientSpaceError)
+	c.Check(diskSpaceErr.Path, Equals, filepath.Join(dirs.GlobalRootDir, "/var/lib/snapd"))
+	c.Check(diskSpaceErr.Snaps, DeepEquals, []string{"one", "two"})
+	c.Check(diskSpaceErr.ChangeKind, Equals, "remove")
+}
+
+func (s *snapmgrTestSuite) TestRemoveManyDiskSpaceCheckDisabled(c *C) {
+	featureFlag := false
+	automaticSnapshot := true
+	failingDiskCheck := 3
+	err := s.testRemoveManyDiskSpaceCheck(c, featureFlag, automaticSnapshot, failingDiskCheck)
+	c.Assert(err, IsNil)
+}
+
+func (s *snapmgrTestSuite) TestRemoveManyDiskSpaceSnapshotDisabled(c *C) {
+	featureFlag := true
+	automaticSnapshot := false
+	failingDiskCheck := 3
+	err := s.testRemoveManyDiskSpaceCheck(c, featureFlag, automaticSnapshot, failingDiskCheck)
+	c.Assert(err, IsNil)
+}
+
+func (s *snapmgrTestSuite) TestRemoveManyDiskSpaceErrorOnSecondSnap(c *C) {
```

Comment:
An explicit test that checks if MockOsutilCheckFreeSpace was called but passes could be added here too.